hpke.go

```diff
@@ -70,21 +75,95 @@ func logVal(name string, value []byte) {
 ///////
 // Core
 
-func setupCore(suite CipherSuite, mode HPKEMode, secret, kemContext, info []byte) (key, nonce []byte) {
-	// context = ciphersuite + mode +
-	//					 len(kemContext) + kemContext +
-	//					 len(info) + info
-	context := []byte{byte(suite.ID) >> 8, byte(suite.ID), byte(mode)}
-	context = append(context, byte(len(kemContext)))
-	context = append(context, kemContext...)
-	context = append(context, byte(len(info)))
-	context = append(context, info...)
-
-	// key = Expand(secret, "hpke key" || context, Nk)
+/*
+
+def EncryptionContext(mode, pkRm, zz, enc, info, psk, pskID, pkIm):
+  VerifyMode(mode, psk, pskID, pkI)
+
+  pkRm = Marshal(pkR)
+  context = concat(mode, ciphersuite, enc, pkRm, pkIm,
+                   len(pskID), pskID, len(info), info)
+
+  secret = Extract(psk, zz)
+  key = Expand(secret, concat("hpke key", context), Nk)
+  nonce = Expand(secret, concat("hpke nonce", context), Nn)
+  return Context(key, nonce)
+
+*/
+
+func defaultPKIm(suite CipherSuite) []byte {
+	return bytes.Repeat([]byte{0x00}, suite.KEM.PublicKeySize())
+}
+
+func defaultPSK(suite CipherSuite) []byte {
+	return bytes.Repeat([]byte{0x00}, suite.KDF.OutputSize())
+}
+
+func defaultPSKID(suite CipherSuite) []byte {
+	return []byte{}
+}
+
+func verifyMode(suite CipherSuite, mode HPKEMode, psk, pskID, pkIm []byte) error {
+	defaultPKIm := defaultPKIm(suite)
+	defaultPSK := defaultPSK(suite)
+	defaultPSKID := defaultPSKID(suite)
+
+	gotPKIm := !bytes.Equal(pkIm, defaultPKIm)
+	noPKIm := bytes.Equal(pkIm, defaultPKIm)
+	gotPSK := !bytes.Equal(psk, defaultPSK) && !bytes.Equal(pskID, defaultPSKID)
+	noPSK := bytes.Equal(psk, defaultPSK) && bytes.Equal(pskID, defaultPSKID)
+
+	ok := false
+	switch mode {
+	case modeBase:
+		ok = noPKIm && noPSK
+	case modePSK:
+		ok = noPKIm && gotPSK
+	case modeAuth:
+		ok = gotPKIm && noPSK
+	case modePSKAuth:
+		ok = gotPKIm && gotPSK
+	}
+
+	if !ok {
+		return fmt.Errorf("Invalid configuration [%d] [%v] [%v]", mode, gotPKIm, gotPSK)
+	}
+
+	return nil
+}
+
+type hpkeContext struct {
+	mode  uint8
+	suite uint16
+	enc   []byte `tls:"head=none"`
+	pkRm  []byte `tls:"head=none"`
+	pkIm  []byte `tls:"head=none"`
+	pskID []byte `tls:"head=2"`
+	info  []byte `tls:"head=2"`
+}
+
+func encryptionContext(suite CipherSuite, mode HPKEMode, pkR KEMPublicKey, zz, enc, info, psk, pskID, pkIm []byte) (key, nonce []byte, err error) {
+	err = verifyMode(suite, mode, psk, pskID, pkIm)
+	if err != nil {
+		return
+	}
+
+	pkRm := suite.KEM.Marshal(pkR)
+
+	contextStruct := hpkeContext{uint8(mode), suite.ID, enc, pkRm, pkIm, pskID, info}
```

Comment:
Should we encode the lengths as we do in the spec?

Reply:
The lengths are encoded, for the variable length things.  That's what the TLS syntax does.